src/multiexp.rs

```diff
@@ -0,0 +1,278 @@
+use crate::curve_jac_ext::CurveJacExt;
+use group::{ff::PrimeField, Group as _};
+use rayon::{
+    prelude::{IndexedParallelIterator, ParallelIterator},
+    slice::ParallelSlice,
+};
+
+/// Performs a multi-exponentiation operation.
+///
+/// This function will panic if coeffs and bases have a different length.
+///
+/// This will use multithreading if beneficial.
+pub fn best_multiexp<C: CurveJacExt>(
+    coeffs: &[C::Scalar],
+    bases: &[C],
+) -> C::ExtendedJacobianCoordinates {
+    let n = coeffs.len();
+    assert_eq!(n, bases.len());
+
+    let num_threads = num_cpus::get();
+    if n > num_threads && n > 32 {
+        let chunk = n / num_threads;
+        let results: Vec<C::ExtendedJacobianCoordinates> = coeffs
```

Comment:
Only the inner buckets result should be collected in Extended Jacobian.

Extended Jacobian mixed addition is faster than projective, but for plain addition, projective is faster.

Reply:
So the signature of best_multiexp should stay with projective coordinates